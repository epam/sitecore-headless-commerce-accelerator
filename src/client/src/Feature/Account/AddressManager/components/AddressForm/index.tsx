//    Copyright 2020 EPAM Systems, Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import * as React from 'react';

import * as Jss from 'Foundation/ReactJss';
import { DependentField, Form, FormValues, Input, Select, Submit } from 'Foundation/ReactJss/Form';

import { Select as PureSelect } from 'Foundation/UI/components/Select';

import { ADDRESS_MANAGER_FIELDS } from './constants';
import { AddressFromProps } from './models';

import './styles.scss';

export default class AddressForm extends React.Component<AddressFromProps> {
  public render() {
    const { countries, defaultValues } = this.props;

    const firstName = defaultValues ? defaultValues.firstName : '';
    const lastName = defaultValues ? defaultValues.lastName : '';
    const address1 = defaultValues ? defaultValues.address1 : '';
    const city = defaultValues ? defaultValues.city : '';
    const zipPostalCode = defaultValues ? defaultValues.zipPostalCode : '';

    return (
      <div className="account-details-form_main active">
        <div className="account-details-form_main_container">
          <div className="form-title">
            <h4>{defaultValues ? 'EDIT' : 'ADD NEW'} ADDRESS</h4>
            <h5>Your Address</h5>
          </div>
          <Form>
            <div className="row">
              <div className="col-lg-12 col-md-12">
                <Jss.Text field={{ value: 'First Name:', editable: 'First Name' }} tag="label" className="required" />
                <Input
                  name={ADDRESS_MANAGER_FIELDS.FIRST_NAME}
                  type="text"
                  required={true}
                  maxLength={100}
                  defaultValue={firstName}
                  fullWidth={true}
                />
              </div>
              <div className="col-lg-12 col-md-12">
                <Jss.Text field={{ value: 'Last Name:', editable: 'Last Name' }} tag="label" className="required" />
                <Input
                  name={ADDRESS_MANAGER_FIELDS.LAST_NAME}
                  type="text"
                  required={true}
                  maxLength={100}
                  defaultValue={lastName}
                  fullWidth={true}
                />
              </div>
              <div className="col-lg-12 col-md-12">
                <Jss.Text
                  field={{ value: 'Address Line:', editable: 'Address Line:' }}
                  tag="label"
                  className="required"
                />
                <Input
                  name={ADDRESS_MANAGER_FIELDS.ADDRESS_LINE}
                  type="text"
                  required={true}
                  maxLength={100}
                  defaultValue={address1}
                  fullWidth={true}
                />
              </div>
              <div className="col-lg-12 col-md-12">
                <Jss.Text field={{ value: 'City:', editable: 'City:' }} tag="label" className="required" />
                <Input
                  name={ADDRESS_MANAGER_FIELDS.CITY}
                  type="text"
                  required={true}
                  maxLength={100}
                  defaultValue={city}
                  fullWidth={true}
                />
              </div>
              <div className="col-lg-6 col-md-6 select-country-state">
                <Jss.Text field={{ value: 'Country:', editable: 'Country' }} tag="label" className="required" />
                <Select
                  fullWidth={true}
                  name={ADDRESS_MANAGER_FIELDS.COUNTRY}
                  required={true}
                  defaultValue={defaultValues.countryCode}
                  className="Select_address_form"
                >
                  <option value="">Not Selected</option>
                  {countries.map((c, index) => (
                    <option key={index} value={c.countryCode}>
                      {c.name}
                    </option>
                  ))}
                </Select>
              </div>
              <div className="col-lg-6 col-md-6 select-country-state">
                <Jss.Text field={{ value: 'State:', editable: 'State' }} tag="label" className="required" />
                <DependentField>
                  {(form) =>
                    form.values[ADDRESS_MANAGER_FIELDS.COUNTRY] ? (
                      <Select
                        fullWidth={true}
                        name={ADDRESS_MANAGER_FIELDS.PROVINCE}
                        required={true}
                        defaultValue={defaultValues.state}
                      >
                        <option value="">Not Selected</option>
                        {this.renderSubdivisions(form.values[ADDRESS_MANAGER_FIELDS.COUNTRY] as string)}
                      </Select>
                    ) : (
                      <PureSelect disabled={true} fullWidth={true}>
                        <option>Not Selected</option>
                      </PureSelect>
                    )
                  }
                </DependentField>
              </div>
              <div className="col-ms-12">
                <Jss.Text field={{ value: 'Postal Code:', editable: 'Postal Code' }} tag="label" className="required" />
                <Input
                  name={ADDRESS_MANAGER_FIELDS.POSTAL_CODE}
                  type="text"
                  required={true}
                  maxLength={100}
                  defaultValue={zipPostalCode}
                  fullWidth={true}
                />
              </div>
            </div>
            <div className="submit-container">
              <Submit
                className="btn btn-outline-main"
                onSubmitHandler={(formValues) => this.submitAddressForm(formValues)}
              >
                Submit
              </Submit>
              <button className="btn btn-outline-focus" onClick={() => this.props.ToggleForm()}>
                Cancel
              </button>
            </div>
          </Form>
        </div>
      </div>
    );
  }

  private submitAddressForm(formValues: FormValues) {
    const { defaultValues, SubmitAction } = this.props;
    const selectedCountry = this.getSelectedCountry(formValues[ADDRESS_MANAGER_FIELDS.COUNTRY] as string);

    const address = {
      ...defaultValues,
      address1: formValues[ADDRESS_MANAGER_FIELDS.ADDRESS_LINE] as string,
      city: formValues[ADDRESS_MANAGER_FIELDS.CITY] as string,
      country: selectedCountry.name,
      countryCode: selectedCountry.countryCode,
      firstName: formValues[ADDRESS_MANAGER_FIELDS.FIRST_NAME] as string,
      isPrimary: false,
      lastName: formValues[ADDRESS_MANAGER_FIELDS.LAST_NAME] as string,
      state: formValues[ADDRESS_MANAGER_FIELDS.PROVINCE] as string,
      zipPostalCode: formValues[ADDRESS_MANAGER_FIELDS.POSTAL_CODE] as string,
    };
    SubmitAction(address);
  }

  private getSelectedCountry(selectedCountryCode: string) {
    const { countries } = this.props;

    return countries.find((c) => c.countryCode === selectedCountryCode);
  }

  private renderSubdivisions(countryCode: string) {
    const selectedCountry = this.getSelectedCountry(countryCode);
    if (!selectedCountry) {
      return null;
    }

    return selectedCountry.subdivisions.map((state, index) => (
      <option key={index} value={state.code}>
        {state.name}
      </option>
    ));
  }
}
