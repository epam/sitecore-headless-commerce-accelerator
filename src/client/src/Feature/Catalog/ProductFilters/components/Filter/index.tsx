//    Copyright 2020 EPAM Systems, Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import classNames from 'classnames';
import * as React from 'react';

import { FilterProps, FilterState } from './models';

export class Filter extends React.Component<FilterProps, FilterState> {
  public constructor(props: FilterProps) {
    super(props);

    this.state = {
      isValuesVisible: true,
    };
  }
  public render() {
    const { first, facet, IsApplied, HandleFacetOnChange } = this.props;
    const { isValuesVisible } = this.state;
    return (
      <div className={classNames({ 'filter-first': first })}>
        <div className="product-filter">
          <div className="product-filter_toggle">
            <h3 className="product-filter_name">{facet.displayName}</h3>
          </div>
          {isValuesVisible && (
            <ul className="product-filter_options options show-all-links">
              {facet.foundValues &&
                facet.foundValues
                  .filter((v) => v.aggregateCount !== 0)
                  .sort((a, b) => {
                    const nameA = a.name.toLowerCase();
                    const nameB = b.name.toLowerCase();
                    if (nameA < nameB) {
                      return -1;
                    }
                    if (nameA > nameB) {
                      return 1;
                    }
                    return 0;
                  })
                  .map((foundValue, i) => {
                    const id = `${foundValue.name}${foundValue.aggregateCount}${i}`;
                    return (
                      <li key={i}>
                        <label className="container">
                          <input
                            type="checkbox"
                            id={id}
                            checked={IsApplied(facet.name, foundValue.name)}
                            onChange={(e) => HandleFacetOnChange(facet.name, foundValue.name, e)}
                          />
                          <span className="checkbox" title={foundValue.name} />
                          <label className="name" htmlFor={id} title={foundValue.name}>{`${foundValue.name}`}</label>
                          <span className="total">{foundValue.aggregateCount}</span>
                        </label>
                      </li>
                    );
                  })}
            </ul>
          )}
          <a
            onClick={(e) => this.toggleFacetVisibility(e)}
            className={classNames('view-all', { 'hide-all': isValuesVisible })}
          >
            <span>
              {isValuesVisible ? (
                <>
                  Hide all
                  <i className="fa fa-angle-up"/>
                </>
              ) : (
                <>
                  Show All
                  <i className="fa fa-angle-down"/>
                </>
              )}
            </span>
          </a>
        </div>
      </div>
    );
  }

  private toggleFacetVisibility(e: React.MouseEvent<HTMLAnchorElement>) {
    e.preventDefault();
    e.stopPropagation();

    this.setState({
      isValuesVisible: !this.state.isValuesVisible,
    });
  }
}
