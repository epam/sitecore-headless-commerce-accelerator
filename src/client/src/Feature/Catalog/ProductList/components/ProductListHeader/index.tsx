//    Copyright 2020 EPAM Systems, Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import * as Jss from 'Foundation/ReactJss';
import * as React from 'react';

import { tryParseUrlSearch } from 'Foundation/Base';

import { FACET_PARAMETER_NAME, facetsManager, PRODUCTS_PER_PAGE } from 'Feature/Catalog/Integration/ProductsSearch';

import { ProductListHeaderProps, ProductListHeaderState } from './models';

import './styles.scss';

export class ProductListHeader extends Jss.SafePureComponent<ProductListHeaderProps, ProductListHeaderState> {
  constructor(props: ProductListHeaderProps) {
    super(props);

    this.state = {
      numberOfDisplayedItems: 0,
    };
  }
  public componentDidUpdate() {
    const { currentPageNumber, itemsCount } = this.props;
    let count = 0;
    if (itemsCount < PRODUCTS_PER_PAGE) {
      this.setState({ numberOfDisplayedItems: itemsCount });
    } else {
      count = (currentPageNumber + 1) * PRODUCTS_PER_PAGE;
      count > itemsCount
        ? this.setState({ numberOfDisplayedItems: itemsCount })
        : this.setState({ numberOfDisplayedItems: count });
    }
  }

  public safeRender() {
    const { search, itemsCount, isLoading } = this.props;
    const { numberOfDisplayedItems } = this.state;
    const parsedSearch = tryParseUrlSearch(search);
    const appliedFacets = facetsManager(parsedSearch[FACET_PARAMETER_NAME]).getFacets();
    return (
      <div className="product_list_header">
        <div className="header">
          {!isLoading && (
            <div className="header_stats">
              <span>
                Showing {numberOfDisplayedItems} of {itemsCount} results
              </span>
            </div>
          )}
        </div>
        <div className="filter-labels">
          {!isLoading &&
            Object.keys(appliedFacets).map((facetName) => {
              const facet = appliedFacets[facetName];
              return facet.map((value, valueIndex) => (
                <div
                  className="filter-label"
                  key={valueIndex}
                  onClick={(e) => this.handleDiscardFacetClick(facetName, value, e)}
                >
                  <i className="fa fa-close" />
                  <span className="filter-label_text">{value}</span>
                </div>
              ));
            })}
        </div>
      </div>
    );
  }

  private handleDiscardFacetClick(name: string, value: string, e: React.MouseEvent<HTMLDivElement>) {
    const { isLoading, search, DiscardFacet } = this.props;
    if (!isLoading) {
      DiscardFacet(name, value, search);
    }
  }
}
