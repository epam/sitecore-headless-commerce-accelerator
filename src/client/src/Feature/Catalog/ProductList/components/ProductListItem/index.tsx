//    Copyright 2020 EPAM Systems, Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import * as React from 'react';

import * as Jss from 'Foundation/ReactJss';
import { NavigationLink } from 'Foundation/UI';

import { ProductRating } from 'Feature/Catalog/ProductRating';
import { ProductListItemProps } from './models';

import './styles.scss';

export class ProductListItem extends Jss.SafePureComponent<ProductListItemProps, Jss.SafePureComponentState> {
  public safeRender() {
    const { product, fallbackImageUrl } = this.props;
    return (
      <figure className="listing-grid-item">
        <div className="img-wrap">
          <NavigationLink to={`/product/${product.productId}`}>
            <img src={!!product.imageUrls[0] ? product.imageUrls[0] : fallbackImageUrl} alt="product image" />
          </NavigationLink>
        </div>
        <figcaption>
          <NavigationLink to={`/product/${product.productId}`}>
            <h3>{product.displayName}</h3>
          </NavigationLink>
          <ProductRating rating={product.customerAverageRating} />
          <div className="price price--adjusted">
            <div className="price_current">
              <span className="price_currency">{product.currencySymbol}</span>
              <span className="price_amount">{product.adjustedPrice.toFixed(2)}</span>
            </div>
            {product.adjustedPrice < product.listPrice && (
              <div>
                <span>&nbsp;-&nbsp; </span>
                <span className="price_full">
                  <span className="price_currency">{product.currencySymbol}</span>
                  <span className="price_amount">{product.listPrice.toFixed(2)}</span>
                </span>
              </div>
            )}
          </div>
          {/* Compare feature commented out
          <input type="checkbox" />
          <label className="compare">Compare</label> */}
        </figcaption>
      </figure>
    );
  }
}
