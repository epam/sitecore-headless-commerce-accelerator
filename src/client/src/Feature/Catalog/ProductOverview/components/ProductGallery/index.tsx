//    Copyright 2020 EPAM Systems, Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import * as React from 'react';

import { LightgalleryItem, LightgalleryProvider } from 'react-lightgallery';

import * as JSS from 'Foundation/ReactJss';

import Swiper, { SwiperRefNode } from 'react-id-swiper';
import { ProductGalleryProps, ProductGalleryState } from './models';
import './styles.scss';

export class ProductGallery extends JSS.SafePureComponent<ProductGalleryProps, ProductGalleryState> {
  private readonly imageSwiperRef: React.MutableRefObject<SwiperRefNode>;
  private readonly carouselSwiperRef: React.MutableRefObject<SwiperRefNode>;
  constructor(props: ProductGalleryProps) {
    super(props);

    this.imageSwiperRef = React.createRef();
    this.carouselSwiperRef = React.createRef();

    this.state = {
    };

    this.goPrev = this.goPrev.bind(this);
    this.goNext = this.goNext.bind(this);
    this.changeIndex = this.changeIndex.bind(this);
  }
  public goPrev() {
    const { images } = this.props;
    const thumbnailSwiper = this.carouselSwiperRef.current.swiper;
    if (thumbnailSwiper) {
      const currentIndex = thumbnailSwiper.activeIndex;
      (currentIndex === 1)
      ? thumbnailSwiper.slideTo(images.length)
      : thumbnailSwiper.slideTo(currentIndex - 1);
    }
  }
  public goNext() {
    const { images } = this.props;
    const thumbnailSwiper = this.carouselSwiperRef.current.swiper;
    if (thumbnailSwiper) {
      const currentIndex = thumbnailSwiper.activeIndex;
      (currentIndex === images.length)
      ? thumbnailSwiper.slideTo(1)
      : thumbnailSwiper.slideTo(currentIndex + 1);
    }
  }

  public changeIndex(e: React.MouseEvent<HTMLDivElement>) {
    const thumbnailSwiper = this.carouselSwiperRef.current.swiper;
    const index = parseInt(e.currentTarget.getAttribute('data-swiper-slide-index'), 10);
    if (thumbnailSwiper) {
      thumbnailSwiper.slideToLoop(index);
    }
  }

  public componentDidUpdate() {
    const gallerySwiper = this.imageSwiperRef.current.swiper;
    const thumbnailSwiper = this.carouselSwiperRef.current.swiper;
    if (gallerySwiper && thumbnailSwiper
    ) {
      gallerySwiper.controller.control = thumbnailSwiper;
      thumbnailSwiper.controller.control = gallerySwiper;
      thumbnailSwiper.height = 140;

      const listOfSlides = thumbnailSwiper.wrapperEl.children;
      for (const slide of listOfSlides as any) {
        slide.addEventListener('click', this.changeIndex);
      }
    }
  }

  protected safeRender() {
    const { images } = this.props;

    const gallerySwiperParams = {
      freeMode: false,
      loop: true,
      loopedSlides: (images && (images.length % 4) === 0 ? 0 : images.length * 3),
      slidesPerView: 1,
      spaceBetween: 10,
    };

    const thumbnailSwiperParams = {
      freeMode: false,
      loop: true,
      loopedSlides: (images && (images.length % 4) === 0 ? 0 : images.length * 3),
      slidesPerView: (images && images.length < 4 ? images.length : 4),
      spaceBetween: 10,
      touchRatio: 0.2,
    };

    return (
      <div className="product-gallery">
        <div className="product-gallery-header">
          <LightgalleryProvider>
            <Swiper {...gallerySwiperParams} effect="fade" ref={this.imageSwiperRef}>
              {images &&
                images.map((src, index) => (
                  <div key={index}>
                    <LightgalleryItem group="any" src={src} thumb={src}>
                      <button>
                        <i className="pe-7s-expand1" />
                      </button>
                    </LightgalleryItem>
                    <div className="single-image">
                      <img src={src} alt="1" className="gallery-image" />
                    </div>
                  </div>
                ))}
            </Swiper>
          </LightgalleryProvider>
        </div>
        <div className="product-gallery-carousel">
          <Swiper {...thumbnailSwiperParams} ref={this.carouselSwiperRef} activeSlideKey="0">
            {images &&
              images.map((src, index) => (
                <div key={index} onClick={(e) => this.changeIndex(e)}>
                  <div className="single-image">
                    <img
                      src={src}
                      className="img-fluid"
                      alt=""
                    />
                  </div>
                </div>
              )
            )}
          </Swiper>
          <button className="product-gallery-carousel-prev" onClick={this.goPrev}>
            <i className="pe-7s-angle-left" />
          </button>
          <button className="product-gallery-carousel-next" onClick={this.goNext}>
            <i className="pe-7s-angle-right" />
          </button>
        </div>
      </div>
    );
  }
}
