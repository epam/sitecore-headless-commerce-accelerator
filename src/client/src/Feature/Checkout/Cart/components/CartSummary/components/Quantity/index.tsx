//    Copyright 2020 EPAM Systems, Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import * as React from 'react';

import * as Jss from 'Foundation/ReactJss';

import { QuantityProductCommon } from 'Feature/Checkout/common/Quantity';
import { ShoppingCart as ShoppingCartApi } from 'Feature/Checkout/Integration/api';
import * as ShoppingCart from 'Feature/Checkout/Integration/ShoppingCart';
import { QuantityProps, QuantityState } from './models';
import './styles.scss';

export class Quantity extends Jss.SafePureComponent<QuantityProps, QuantityState> {
  constructor(props: QuantityProps) {
    super(props);
    const quantity = this.props.cartLine.quantity;
    this.state = {
      quantityString: quantity.toString(),
    };
  }

  public setQuantityString(increase: boolean) {
    const { quantityString } = this.state;
    const { RemoveCartLine, cartLine } = this.props;
    const quantity = parseInt(quantityString, 10);
    const newQuantity = increase ? quantity + 1 : quantity - 1;
    if (!increase && quantity === 1) {
      RemoveCartLine(cartLine);
    } else {
      this.setState({ quantityString: newQuantity.toString() }, () => this.updateCartLine(cartLine));
    }
  }

  public updateCartLine(cartLine: ShoppingCart.ShoppingCartLine) {
    const quantity = parseInt(this.state.quantityString, 10);
    if (quantity >= 0) {
      const updateCartLineModel: ShoppingCartApi.CartItemDto = {
        productId: cartLine.product.productId,
        quantity,
        variantId: cartLine.variant.variantId,
      };
      this.props.UpdateCartLine(updateCartLineModel);
    } else {
      this.setState({
        quantityString: cartLine.quantity.toString(),
      });
    }
  }

  public safeRender() {
    const cartLine = this.props.cartLine;
    const quantityString = this.state.quantityString;
    return (
      <QuantityProductCommon
        cartLineId={cartLine.id}
        setQuantity={(isIncrease: boolean) => this.setQuantityString(isIncrease)}
        quantityString={quantityString}
      />
    );
  }
}
