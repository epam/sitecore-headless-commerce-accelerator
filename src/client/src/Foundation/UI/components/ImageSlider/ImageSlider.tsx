//    Copyright 2020 EPAM Systems, Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import React, { Children, FC, ReactElement, useCallback, useMemo, useRef, useState } from 'react';

import { get } from 'lodash';
import Swiper from 'react-id-swiper';
import { useInterval } from 'react-use';

import { cnImageSlider } from './cn';
import './ImageSlider.scss';

const MIN_SLIDES_FOR_DOTS = 2;

export type ImageSliderProps = {
  children?: ReactElement | ReactElement[];
  delay?: number;
};

export const ImageSlider: FC<ImageSliderProps> = ({ children, delay = 1000 }) => {
  const ref = useRef(null);
  const [swiperState, setSwiperState] = useState<'stop' | 'forward' | 'backward'>('stop');
  const defaultParams = useMemo(() => {
    const params = {
      lazy: true,
    };

    if (Children.count(children) >= MIN_SLIDES_FOR_DOTS) {
      params['pagination'] = {
        bulletActiveClass: 'ImageSlider-SwiperBullets_active',
        el: '.swiper-pagination',
      };
    }

    return params;
  }, [children]);

  useInterval(
    () => {
      const swiper = get(ref, ['current', 'swiper']);

      if (swiperState === 'forward' && swiper) {
        swiper.slideNext();
      }

      if (swiperState === 'backward' && swiper) {
        swiper.slidePrev();
      }
    },
    swiperState === 'stop' ? null : delay,
  );

  const goNext = useCallback(() => {
    setSwiperState('forward');
  }, [setSwiperState]);

  const goPrev = useCallback(() => {
    setSwiperState('backward');
  }, [setSwiperState]);

  const clear = useCallback(() => {
    setSwiperState('stop');
  }, [setSwiperState]);

  return (
    <div className={cnImageSlider()}>
      <div className={cnImageSlider('SwiperContainer')}>
        <Swiper {...defaultParams} ref={ref}>
          {children}
        </Swiper>
      </div>
      <div className={cnImageSlider('SwiperOverlay')} onMouseLeave={clear}>
        <div className={cnImageSlider('ButtonLeft')} onMouseEnter={goPrev} />
        <div className={cnImageSlider('ButtonRight')} onMouseEnter={goNext} />
      </div>
    </div>
  );
};
