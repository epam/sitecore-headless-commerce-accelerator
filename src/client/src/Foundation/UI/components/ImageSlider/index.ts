export * from './ImageSlider';
export * from './Slide';
