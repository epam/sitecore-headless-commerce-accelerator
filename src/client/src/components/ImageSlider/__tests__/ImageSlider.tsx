//    Copyright 2021 EPAM Systems, Inc.
//
//    Licensed under the Apache License, Version 2.0 (the "License");
//    you may not use this file except in compliance with the License.
//    You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//    Unless required by applicable law or agreed to in writing, software
//    distributed under the License is distributed on an "AS IS" BASIS,
//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//    See the License for the specific language governing permissions and
//    limitations under the License.

import React from 'react';
import renderer from 'react-test-renderer';

import { ImageSlider, Slide } from 'components/ImageSlider';

const fallbackImageUrl = '/-/media/HCA/Project/HCA/no-image250x150.ashx';

it('ImageSlider renders correctly', () => {
  const tree = renderer
    .create(
      <ImageSlider>
        <Slide url={fallbackImageUrl} />
      </ImageSlider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
