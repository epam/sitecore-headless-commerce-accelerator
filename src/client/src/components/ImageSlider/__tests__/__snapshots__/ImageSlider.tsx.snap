// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageSlider renders correctly 1`] = `
<div
  className="ImageSlider"
>
  <div
    className="ImageSlider-SwiperContainer"
  >
    <div
      className="swiper-container"
    >
      <div
        className="swiper-wrapper"
      >
        <div
          className="ImageSlider-Slide swiper-slide"
        >
          <img
            alt=""
            className="ImageSlider-Image swiper-lazy"
            src="/-/media/HCA/Project/HCA/no-image250x150.ashx"
          />
          <div
            className="swiper-lazy-preloader swiper-lazy-preloader-white"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="ImageSlider-SwiperOverlay"
    onMouseLeave={[Function]}
  >
    <div
      className="ImageSlider-ButtonLeft"
      onMouseEnter={[Function]}
    />
    <div
      className="ImageSlider-ButtonRight"
      onMouseEnter={[Function]}
    />
  </div>
</div>
`;
